README
								-*-text-*-

Turtle -- a constraint-imperative programming language
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This is version 1.0.0 of Turtle, an implementation of the constraint
imperative programming language Turtle.

Turtle versions with an odd middle number, i.e. 0.1.* are unstable
development versions.  Even middle numbers indicate stable versions.

The next stable release will be version 1.2.0.

Please send bug reports to <mgrabmue@cs.tu-berlin.de>


Turtle Installation ==================================================

  Please see the file INSTALL in the source distribution for details
  on how to build and install the Turtle system.


Turtle Documentation =================================================

  The doc directory contains the documentation for the Turtle system
  in Texinfo format.

  The Turtle WWW page can be found at

  http://user.cs.tu-berlin.de/~mgrabmue/turtle


Turtle License =======================================================

  Turtle is free software; you can redistribute it and/or modify it
  under the terms of the GNU General Public License as published by the
  Free Software Foundation; either version 2, or (at your option) any
  later version.

  Turtle is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this package; see the file COPYING.  If not, write to the
  Free Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,
  MA 02111-1307, USA.


About This Distribution ==============================================

Interesting files include:

  AUTHORS       Names of authors and contributors.

  COPYING       The license under which Turtle is distributed.

  HACKING       Informations for hacking on Turtle.

  INSTALL       This file contains instructions on building and 
                installing Turtle.

  NEWS          Describes user-visible changes since the last release 
                of Turtle.

Files are usually installed according to the prefix specified to
configure, /usr/local by default.  Building and installing gives you:

Executables, in ${prefix}/bin:

  turtle        A standalone compiler for Turtle.  Type 
                `turtle --help' to get an overview of the available
		command line options.


Libraries, in ${prefix}/lib.  Depending on the platform and options
given to configure, you may get shared libraries in addition or
instead of these static libraries:

  libturtle.a   An object library containing the Turtle compiler,
	        You can use Turtle in your own programs by linking 
                against this.

  libturtle.la  libtool archive of the above.

  liburtle.so,
  libturtle.so.0,
  libturtle.so.0.0.0    
                Shared library version of the above.

  libturtlert.a This object library contains the Turtle runtime 
                system.  It is linked against programs created by
                the Turtle compiler.

  libturtlert.la
                libtool archive of the above.

  libturtlert.so,
  libturtlert.so.0,
  libturtlert.so.0.0.0
                Shared library version of the runtime system.

  turtle/1.0.0  This directory contains the modules of the Turtle
                standard library.  *.h files contain the C-level
                interface, *.o files contain the compiled object code
	        and *.ifc contain the interface definitions, to be
	        used by the Turtle compiler.


Header files, in ${prefix}/include:

  libturtle/libturtle.h  For including when using the Turtle library.
  libturtle/*.h          Various header files, included by the above.


Documentation in Info format, in ${prefix}/info:

  turtle.info,
  turtle.info-1,
  turtle.info-2,
  turtle.info-3,
  turtle.info-4 The Turtle reference manual.


The Turtle source tree is laid out as follows:

  crawl         The run-time support and some modules.

  doc           Documentation (see above).

  emacs         The GNU Emacs Turtle mode, for easier editing of 
		Turtle source code.  Maybe it works with XEmacs, too.

  examples      Some example Turtle programs, for testing the compiler
                and for demonstrating the language.

  libturtle     The Turtle library source code.

  tests         The Turtle test suite.

  tools         Some (experimental) tools, written in Turtle. 

  turtle        The stand-alone compiler.


Obtaining Turtle =====================================================

  The latest official Turtle release is available via HTTP from

  http://user.cs.tu-berlin.de/~mgrabmue/turtle/download/turtle-1.0.0.tar.gz


NEWS
Turtle NEWS:  History of user-visible changes.

Copyright (C) 2003 Martin Grabmueller

Please send Turtle bug reports to <mgrabmue@cs.tu-berlin.de>


Changes between 0.1.7 and 1.0.0:

This is the first release of Turtle.


Changes between 0.1.6 and 0.1.7:

Cleanup work on Diploma Thesis.


Changes between 0.1.5 and 0.1.6:

Documentation improvements, the examples work again.


Changes between 0.1.4 and 0.1.5:

Major overhaul of constrainable variables, constraint compilation
etc.


Changes between 0.1.3 and 0.1.4:

Mainly work on the docs.

Tiny improvements to the web server.


Changes between 0.1.2 and 0.1.3:

Mainly work on the docs.


Changes between 0.1.1 and 0.1.2:

More work on the constraint-solving part of the implementation.

Various fixes, example additions.


Changes between 0.1.0 and 0.1.1:

Nearly everything changed.  We now have:

- a working compiler for the imperative subset of Turtle,

- some useful modules in the standard library and

- some documentation for the compiler, the standard library and
  the language.


Changes between 0.0.0 and 0.1.0:

The whole package is new.


Local variables:
mode: outline
paragraph-separate: "[ 	]*$"
end:


turtle/turtle.c
/* turtle/turtle.c - Turtle compiler.
 
  Copyright (C) 2003 Martin Grabmueller <mgrabmue@cs.tu-berlin.de>
 
  This is free software; you can redistribute it and/or modify it
  under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2, or (at your option)
  any later version.
  
  This software is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with this package; see the file COPYING.  If not, write to the
  Free Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,
  MA 02111-1307, USA.  */

#if HAVE_CONFIG_H
# include <config.h>
#endif 

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#if HAVE_GETOPT_H
# include <getopt.h>
#endif

#include <version.h>

#include <libturtle/libturtle.h>


static char program_name[] = "turtle";

static char version_string[] = __turtle_version;

static char system_init_file_name[] = "init";
static char user_init_file_name[] = ".turtle";

static char * program_path = NULL;


/* Defining here the struct and #define's for getopt_long() if it is
   in libiberty.a but could not be found in getopt.h

   [This replacement for getopt_long() is due to Stefan Jahn
    <ela@lkcc.org>] 
 */
#if defined(HAVE_GETOPT_LONG) && !defined(DECLARED_GETOPT_LONG)

struct option
{
  const char *name;
  int has_arg;
  int *flag;
  int val;
};

#define no_argument       0
#define required_argument 1
#define optional_argument 2

extern int getopt_long(int argc, 
		       char * const argv[], 
		       const char *optstring,
		       const struct option *longopts, 
		       int *longindex);

extern int optind;
extern char * optarg;
#endif /* DECLARED_GETOPT_LONG */

#ifndef HAVE_GETOPT
char *optarg = NULL;
int optind = 1;
int opterr = 0;
int optopt = 0;

int 
getopt (int argc, char * const argv[], const char *optstring) 
{
  static int current_arg = 1, current_opt = 0;
  int n;

  while (current_arg < argc) 
    {
      if (current_arg != 0 || argv[current_arg][0] == '-') 
        {
          if (current_opt == 0)
            current_opt = 1;
          while (argv[current_arg][current_opt] != '\0')
	    {
	      n = 0;
	      /* go through all option characters */
	      while (optstring[n]) 
		{
		  if (optstring[n] == argv[current_arg][current_opt])
		    {
		      current_opt++;
		      if (optstring[n + 1] == ':') 
			optarg = argv[current_arg + 1];
		      else
			optarg = NULL;
		      if (argv[current_arg][current_opt] == '\0')
			{ 
			  current_arg++;
			  current_opt = 0;
			} 
		      optind = current_arg;
		      return optstring[n];
		    }
		  n++;
		}
            }
	  current_opt++;
	}
      current_arg++;
    }

  current_arg = 1;
  current_opt = 0;
  return EOF;
}
#endif /* not HAVE_GETOPT */


/* Print a usage help message to standard output.  */
static void 
usage (void)
{
#if HAVE_GETOPT_LONG
  printf ("usage: %s [OPTION...] FILENAME...\n\
  -h, --help                 display this help and exit\n\
  -v, --version              display version information and exit\n\
  -V, --verbose              increase verbosity level\n\
  -m, --main=NAME            compile a main program and output as NAME\n\
  -p, --module-path=PATH     set the search path for modules\n\
  -I, --include-path=PATH    set the search path for runtime header files\n\
  -L, --library-path=PATH    set the search path for library files\n\
  -z, --pragma=PRAGMA        set compilation pragma\n\
    where PRAGMA is one of\n\
      handcoded              handcoded module\n\
      foreign                foreign expressions\n\
      compile-only           do not run the C compiler\n\
      static                 produce a static executable\n\
      turtledoc              generate Texinfo documentation from comments\n\
      deps                   write dependency information to .P file\n\
      deps-stdout            write dependency information to standard output\n\
  -O, --optimize=FLAGS       set optimization flags\n\
    where FLAGS is one or more of\n\
      C                      optimize module-local calls\n\
      c                      do not optimize module-local calls\n\
      J                      convert module-local calls to jumps\n\
      j                      do not convert module-local calls to jumps\n\
      G                      optimize GC checks over basic blocks\n\
      g                      do not optimize GC checks over basic blocks\n\
      D                      inline data constructors etc.\n\
      d                      do not inline data constructors etc.\n\
      0-6                    set optimization level for C compiler\n\
  -d, --debug=MODIFIER       set debugging options\n\
    where MODIFIER is one or more of\n\
      a                      print AST after parsing\n\
      e                      print top-level environment of module\n\
      i                      dump the HIL representation to a .il file\n\
      b                      print bytecode after code generation\n\
Report bugs to mgrabmue@cs.tu-berlin.de\n",
	  program_name);
#else /* !HAVE_GETOPT_LONG */
  printf ("usage: %s [OPTION...] FILENAME...\n\
  -h                 display this help and exit\n\
  -v                 display version information and exit\n\
  -V                 increase verbosity level\n\
  -m NAME            compile a main program and output as NAME\n\
  -p PATH            set the search path for modules\n\
  -I PATH            set the search path for runtime header files\n\
  -L PATH            set the search path for library files\n\
  -z PRAGMA          set compilation pragma\n\
    where PRAGMA is one of\n\
      handcoded      handcoded module\n\
      foreign        foreign expressions\n\
      compile-only   do not run the C compiler\n\
      static         produce a static executable\n\
      turtledoc      generate Texinfo documentation from comments\n\
      deps           write dependency information to .P file\n\
      deps-stdout    write dependency information to standard output\n\
  -O FLAGS           set optimization flags\n\
    where FLAGS is one or more of\n\
      C              optimize module-local calls\n\
      c              do not optimize module-local calls\n\
      J              convert module-local calls to jumps\n\
      j              do not convert module-local calls to jumps\n\
      G              optimize GC checks over basic blocks\n\
      g              do not optimize GC checks over basic blocks\n\
      D              inline data constructors etc.\n\
      d              do not inline data constructors etc.\n\
      0-6            set optimization level for C compiler\n\
  -d MODIFIER        set debugging options\n\
    where MODIFIER is one or more of the letters\n\
      a              print AST after parsing\n\
      e              print top-level environment of module\n\
      i              dump the HIL representation to a .il file\n\
      b              print bytecode after code generation\n\
Report bugs to mgrabmue@cs.tu-berlin.de\n",
	  program_name);
#endif /* !HAVE_GETOPT_LONG */
}


/* Print version information to standard output.  */
static void 
version (void)
{
  printf ("%s %s\n", program_name, version_string);
}


int
main (int argc, char * argv[])
{
  int arg;
  int index;
  int exit_code = 0;
  char home_init_file[12];

  struct ttl_compile_options options;

#if HAVE_GETOPT_LONG
  static struct option command_line_options[] =
  {
    {"help", no_argument, NULL, 'h'},
    {"version", no_argument, NULL, 'v'},
    {"verbose", no_argument, NULL, 'V'},
    {"debug", required_argument, NULL, 'd'},
    {"pragma", required_argument, NULL, 'z'},
    {"optimize", required_argument, NULL, 'O'},
    {"main", required_argument, NULL, 'm'},
    {"module-path", required_argument, NULL, 'p'},
    {"include-path", required_argument, NULL, 'I'},
    {"library-path", required_argument, NULL, 'L'},
    {NULL, 0, NULL, 0}
  };
#endif /* HAVE_GETOPT_LONG */

  program_path = argv[0];


  ttl_init_compile_options (&options);

#if HAVE_GETOPT_LONG
  while ((arg = getopt_long (argc, argv, "+hvVm:d:I:L:z:O:p:",
			     command_line_options, &index)) != EOF)
#else /* !HAVE_GETOPT_LONG */
  while ((arg = getopt (argc, argv, "hvVm:d:I:L:z:O:p:")) != EOF)
#endif /* !HAVE_GETOPT_LONG */
    {
      switch (arg)
	{
	case 'h':
	  usage ();
	  exit (0);
	  break;

	case 'v':
	  version ();
	  exit (0);
	  break;

	case 'V':
	  options.verbose++;;
	  break;

	case 'm':
	  options.main = 1;
	  options.program_name = optarg;
	  break;

	case 'p':
	  options.module_path = optarg;
	  break;

	case 'I':
	  options.include_path = optarg;
	  break;

	case 'L':
	  options.library_path = optarg;
	  break;

	case 'd':
	  {
	    char * p = optarg;
	    while (*p)
	      {
		switch (*p)
		  {
		  case 'e':
		    options.debug_dump_env = 1;
		    break;
		  case 'a':
		    options.debug_dump_ast = 1;
		    break;
		  case 'i':
		    options.debug_dump_il_file = 1;
		    break;
		  case 'b':
		    options.debug_dump_bytecode = 1;
		    break;
		  }
		p++;
	      }
	    break;
	  }

	case 'z':
	  {
	    if (!strcmp (optarg, "handcoded"))
	      options.pragma_handcoded = 1;
	    else if (!strcmp (optarg, "foreign"))
	      options.pragma_foreign = 1;
	    else if (!strcmp (optarg, "compile-only"))
	      options.pragma_compile_only = 1;
	    else if (!strcmp (optarg, "turtledoc"))
	      options.pragma_turtledoc = 1;
	    else if (!strcmp (optarg, "deps"))
	      options.pragma_printdeps = 1;
	    else if (!strcmp (optarg, "deps-stdout"))
	      options.pragma_printdepsstdout = 1;
	    else if (!strcmp (optarg, "static"))
	      options.link_static = 1;
	    else
	      {
		fprintf (stderr, "turtle: invalid pragma: %s\n", optarg);
		usage ();
		exit (1);
	      }
	    break;
	  }

	case 'O':
	  {
	    char * p = optarg;
	    while (*p)
	      {
		switch (*p)
		  {
		  case 'C':
		    options.opt_local_calls = 1;
		    break;
		  case 'c':
		    options.opt_local_calls = 0;
		    break;
		  case 'J':
		    options.opt_local_jumps = 1;
		    break;
		  case 'j':
		    options.opt_local_jumps = 0;
		    break;
		  case 'G':
		    options.opt_merge_gc_checks = 1;
		    break;
		  case 'g':
		    options.opt_merge_gc_checks = 0;
		    break;
		  case 'D':
		    options.opt_inline_constructors = 1;
		    break;
		  case 'd':
		    options.opt_inline_constructors = 0;
		    break;
		  case '0':
		  case '1':
		  case '2':
		  case '3':
		  case '4':
		  case '5':
		  case '6':
		    options.opt_gcc_level = *p - '0';
		    break;
		  }
		p++;
	      }
	    break;
	  }

	default:
	  usage ();
	  exit (1);
	  break;
	}
    }
 parsing_end:

  if (optind >= argc)
    {
      printf ("%s: No input files.\n", program_name);
      exit (1);
    }

  if (options.main && argc - optind > 1)
    {
      printf ("%s: Only one file can be compiled with --main=NAME\n",
	      program_name);
      exit (1);
    }

  {
    int i;

    i = optind;
    while (argv[i])
      {
	if (ttl_compile (argv[i], &options) > 0)
	  exit_code = 1;
	i++;
      }
  }

  if (ttl_allocated_bytes != 0)
    fprintf (stderr, "%d bytes leaked.\n", ttl_allocated_bytes);

  return exit_code;
}

/* End of turtle.c.  */
